### ISSUE TITLE
Crashed when I train my datasets

### ISSUE THREAD

**Issue by yyfsss** [tiny-dnn/tiny-dnn#373]
My dataset is so small and I add the labels to the dataset. But when I train my dataset. It always says that "Vector iterator + offset out of range" . I am almost crazy because my teacher ask me to show the results recently.
here is the train cpp.

**Comment by yyfsss** [comment]
#include <iostream>
#include <memory>
#include "tiny_dnn/tiny_dnn.h"
#include <opencv2/opencv.hpp>
#include <strstream>

#include <opencv2/imgproc.hpp>
#include <boost/foreach.hpp>
#include <boost/filesystem.hpp>
using namespace boost::filesystem;
using namespace tiny_dnn;
using namespace tiny_dnn::activation;
using namespace cv;
using namespace tiny_dnn::layers;
#include<strstream>
//**********************************************************************************//  
 
const int  TRNUM = 15;   
const int  TENUM = 6;      
const int  C = 3;      
using namespace std;
//*********************************************************************************//
//
void sample1_convnet(const string& data_dir_path = "../../data");


///////////////////////////////////////////////////////////////////////////////
// learning convolutional neural networks (LeNet-5 like architecture)

void sample1_convnet(const string& data_dir_path) {
	// construct LeNet-5 architecture
	network<sequential> nn;
	adagrad optimizer;

	// connection table [Y.Lecun, 1998 Table.1]
#define O true
#define X false
	static const bool connection[] = {
		O, X, X, X, O, O, O, X, X, O, O, O, O, X, O, O,
		O, O, X, X, X, O, O, O, X, X, O, O, O, O, X, O,
		O, O, O, X, X, X, O, O, O, X, X, O, X, O, O, O,
		X, O, O, O, X, X, O, O, O, O, X, X, O, X, O, O,
		X, X, O, O, O, X, X, O, O, O, O, X, O, O, X, O,
		X, X, X, O, O, O, X, X, O, O, O, O, X, O, O, O
	};
#undef O
#undef X

	nn << convolutional_layer<tan_h>(32, 32, 5, 1, 6) // 32x32 in, 5x5 kernel, 1-6 fmaps conv
		<< average_pooling_layer<tan_h>(28, 28, 6, 2) // 28x28 in, 6 fmaps, 2x2 subsampling
		<< convolutional_layer<tan_h>(14, 14, 5, 6, 16,
		connection_table(connection, 6, 16)) // with connection-table
		<< average_pooling_layer<tan_h>(10, 10, 16, 2)
		<< convolutional_layer<tan_h>(5, 5, 5, 16, 120)
		<< fully_connected_layer<tan_h>(120, C);
}
//**************************`**********************************************************************
// convert image to vec_t  

void convert_image(const std::string& filename,
	std::vector<vec_t> &data,
	float min,
	float max,
	int padding){
	Mat src = imread(filename, -1);
	int W = src.cols;
	int H = src.rows;
	float denominator = 255;
	vec_t img(3 * (H + 2 * padding)*(W + 2 * padding), min);

	for (int i = 0; i < H; i++) { // Go over all rows  
		for (int j = 0; j < W; j++) { // Go over all columns  
			for (int c = 0; c < 3; c++) { // Go through all channels  
				img[W*H*c + W*(i + padding) + (j + padding)] = src.at<cv::Vec3b>(i, j)[c] / denominator*(max - min) + min;
			}
		}
	}
	data.push_back(img);
}

	
	string int2string(int&i){
		strstream ss;  string str;
		ss << i; ss >> str;
		return str;
	}
//*****************************************************************************************
	int main() {
		// specify loss-function and learning strategy  
		std::cout << "load models..." << std::endl;
		network<sequential> nn;
		adagrad optimizer;
		sample1_convnet();
		//**********************************************************************************//  
		  
		std::vector<label_t> train_labels, test_labels;
		std::vector<vec_t> train_images, test_images;
		for (int i = 1; i <= TRNUM; i++){
			string  str0 = int2string(i);
			
			string trainpath = "E://train2/"+str0+".jpg";  
			cout << trainpath  << endl;
				
			  
			convert_image(trainpath, train_images ,-1, 1, 2);
		}
		cout << "cout << train_images.size():" << train_images.size() << endl;
		  
		for (int i = 1; i <= TENUM; i++){
			string  str1 = int2string(i);

			string testpath = "E://test2/"+str1+".jpg";    
			convert_image(testpath, test_images, -1, 1, 2);
		}
		cout << "cout << test_images.size():" << test_images.size() << endl;
		//**********************************************************************************//  
		  
		printf("加载训练集标签\n");//load the trainning tables
		int Ci;   
		for (size_t k = 1; k <= C; k++){    
			printf("第%d类样本的个数：", k);  scanf("%d", &Ci);
			for (size_t j = 1; j <= Ci; j++){
				train_labels.push_back((label_t)k);
			}
		}
		cout << "train_labels" << train_labels.size() << endl;
		printf("加载测试集标签\n");
		int Ti;  
		for (size_t k = 1; k <=C; k++){    
			printf("第%d类样本的个数：",k);  scanf("%d", &Ti);
			for (size_t j = 1; j <= Ti; j++){
				test_labels.push_back((label_t)k);
			}
		}
		cout << "test_labels" << test_labels.size() << endl;
		  
		std::cout << "start training" << std::endl;

		progress_display disp(train_images.size());  
		timer t;
		int minibatch_size = 10;
		int num_epochs = 30;        
        optimizer.alpha *= std::sqrt(minibatch_size);
		
		  
		auto on_enumerate_epoch = [&](){
			std::cout << t.elapsed() << "s elapsed." << std::endl;
			tiny_dnn::result res = nn.test(test_images, test_labels);
			std::cout << res.num_success << "/" << res.num_total << std::endl;

			disp.restart(train_images.size());
			t.restart();
		};

		auto on_enumerate_minibatch = [&](){
			disp += minibatch_size;
		};

		  
		nn.train<mse>(optimizer, train_images, train_labels, minibatch_size, 30, on_enumerate_minibatch, on_enumerate_epoch);

		std::cout << "end training." << std::endl;

		  
		nn.test(test_images, test_labels).print_detail(std::cout);

		  
		std::ofstream ofs("LeNet-weights");
		ofs << nn;

		system("pause");
	}

**Comment by bhack** [comment]
Can you move this on stackoverflow using tiny-dnn tag?

**Comment by Randl** [comment]
@yyfsss If you still have the problem please post the question on stackoverflow as asked and you may post the link to the question here.
